syntax = "proto3";

package exchange;

enum Side { BUY = 0; SELL = 1; }

message Order {
  uint64  id        = 1;
  uint64  user_id   = 2;
  Side    side      = 3;
  double  price     = 4;   // NaN -> market order
  uint32  quantity  = 5;
}

message SubmitReply { repeated Trade trades = 1; }
message CancelReply { bool ok = 1; }
message WalletReply { double usd = 1; double asset = 2; }

message Trade {
  uint64 buy_id  = 1;
  uint64 sell_id = 2;
  double price   = 3;
  uint32 quantity= 4;
  double timestamp = 5;
}

service Exchange {
  rpc SubmitOrder (Order)  returns (SubmitReply);
  rpc CancelOrder (Order)  returns (CancelReply);  // only id+user_id set
  rpc GetWallet   (Order)  returns (WalletReply);  // only user_id set
}
